File: Dart/Sort/Comb_Sort.dart
/*The Comb Sort is a variant of the Bubble Sort. Like the Shell sort, the Comb Sort increases the gap used in comparisons and exchanges.
*/

/*
 Input
array = [1, 451, 562, 2, 99, 78, 5]

// Ouput
SortedArray = [1, 2, 5, 78, 99, 451, 562]
Time complexities:
	Worst Case :O(n2) 
  Best Case :O(nlogn)   
  Average Case:Ω(N2/2p)
Space Complexity: Auxilary Space:O(1)  
*/

// function for combsort
void combSort(List list) {
  int gpVal = list.length;
  double shrink = 1.3;
  bool sortedBool = false;

  while (!sortedBool) {
    gpVal = (gpVal / shrink).floor();
    if (gpVal > 1) {
      sortedBool = false;
    } else {
      gpVal = 1;
      sortedBool = true;
    }

    int i = 0;
    while (i + gpVal < list.length) {
      if (list[i] > list[i + gpVal]) {
        swap(list, i, gpVal);
        sortedBool = false;
      }
      i++;
    }
  }
}

// function to swap the values
void swap(List list, int i, int gpVal) {
  int temp = list[i];
  list[i] = list[i + gpVal];
  list[i + gpVal] = temp;
}

void main() {
  //Get the dummy array
  List arr = [1, 451, 562, 2, 99, 78, 5];
  // applying combSort function
  combSort(arr);
  // printing the sortedBool value
  print(arr);
}

File: Python/sort/comb_sort.py
"""
CombSort Algorithm Implementation in Python 3

Purpose:
Comb sort is a relatively simple sorting algorithm.
Comb sort improves on bubble sort algorithm.
In bubble sort, distance (or gap) between two compared elements is always one.
Comb sort improvement is that gap can be much more than 1, in order to prevent
slowing down by small values at the end of a list.

Time Complexity :
Worst case complexity is O(n^2)
Best Case complexity is O(n log n)

Space Complexity: O(1)

Argument/ Return Type:
arr : Unorded list
output : Return list in ascending order

"""


def combsort(arr, n):

    """
    Pure implementation of comb sort algorithm in Python
    :param data: mutable collection with comparable items
    :return: the same collection in ascending order
    Examples:
    >>> comb_sort([0, 5, 3, 2, 2])
    [0, 2, 2, 3, 5]
    >>> comb_sort([])
    []
    >>> comb_sort([99, 45, -7, 8, 2, 0, -15, 3])
    [-15, -7, 0, 2, 3, 8, 45, 99]

    """

    gap = n
    while gap > 1:
        gap = int(gap / 1.3)
        for i in range(0, n - gap):
            if arr[i] > arr[i + gap]:
                arr[i], arr[i + gap] = arr[i + gap], arr[i]
    return arr


a = int(input("Enter the number of elements in the array: "))
li = []
for i in range(0, a):
    li.append(int(input('Enter the element one at a time: ')))
print('The unsorted array you have given is:', li)
print('The sorted array using combsort method is: ', end="")
print(combsort(li, len(li)))
